logstash-core/lib/logstash/config/mixin.rb

```diff
@@ -99,6 +99,17 @@ def config_init(params)
       params[name.to_s] = deep_replace(value)
     end
 
+    # Intercept codecs that have not been instantiated
+    params.each do |name, value|
+      validator = self.class.validator_find(name)
+      next unless validator && validator[:validate] == :codec && value.kind_of?(String)
+
+      codec_klass = LogStash::Plugin.lookup("codec", value)
+      codec_instance = LogStash::Plugins::Contextualizer.initialize_plugin(execution_context, codec_klass)
+
+      params[name.to_s] = LogStash::Codecs::Delegator.new(codec_instance)
```

Comment:
and these instantiated on demand previously?

Reply:
This was a weird one.

When a pipeline definition included a codec that parsed by the LIR as a plugin definition (e.g., `codec => rubydebug {}`), the codec was instantiated by the plugin factory in the normal way, but when the LIR parsed a codec as a string (quoted or bareword. e.g., `codec => rubydebug`), the instantiation was done by the input/output in `Config::Mixin::validate_value(value, validator)` which as a class method does not have access to the instance's execution context.

It looks like I forgot to include the removal of the string handling there, which should now never happen because we intercept it here.